test/modules/GenericMqttTester/src/message_channel.rs

```diff
@@ -0,0 +1,160 @@
+    PublishHandle, ReceivedPublication,
+};
+
+use crate::{MessageTesterError, BACKWARDS_TOPIC};
+
+#[derive(Debug, Clone)]
+pub struct MessageHandlerShutdownHandle(Sender<()>);
+
+impl MessageHandlerShutdownHandle {
+    pub fn new(sender: Sender<()>) -> Self {
+        Self(sender)
+    }
+
+    pub async fn shutdown(mut self) -> Result<(), MessageTesterError> {
+        self.0
+            .send(())
+            .await
+            .map_err(MessageTesterError::SendShutdownSignal)
+    }
+}
+
+/// Responsible for receiving publications and taking some action.
+#[async_trait]
+pub trait MessageHandler {
+    /// Starts handling messages sent to the handler
+    async fn handle(&mut self, publication: ReceivedPublication) -> Result<(), MessageTesterError>;
+}
+
+/// Responsible for receiving publications and reporting result to the Test Result Coordinator.
+pub struct ReportResultMessageHandler {}
+
+impl ReportResultMessageHandler {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+#[async_trait]
+impl MessageHandler for ReportResultMessageHandler {
+    async fn handle(
+        &mut self,
+        received_publication: ReceivedPublication,
+    ) -> Result<(), MessageTesterError> {
+        info!("should send result to TRC, but not implemented yet");
+        Ok(())
+    }
+}
+
+/// Responsible for receiving publications and sending them back to the downstream edge.
+pub struct RelayingMessageHandler {
+    publish_handle: PublishHandle,
+}
+
+impl RelayingMessageHandler {
+    pub fn new(publish_handle: PublishHandle) -> Self {
+        Self { publish_handle }
+    }
+}
+
+#[async_trait]
+impl MessageHandler for RelayingMessageHandler {
+    async fn handle(
+        &mut self,
+        received_publication: ReceivedPublication,
+    ) -> Result<(), MessageTesterError> {
+        info!("relaying publication {:?}", received_publication);
+
+        let new_publication = Publication {
+            topic_name: BACKWARDS_TOPIC.to_string(),
+            qos: QoS::ExactlyOnce,
+            retain: true,
+            payload: received_publication.payload,
+        };
+        self.publish_handle
+            .publish(new_publication)
+            .await
+            .map_err(MessageTesterError::Publish)?;
+
+        Ok(())
+    }
+}
+
+/// Serves as a channel between a mqtt client's received publications and a message handler.
+/// Exposes a message channel to send messages to a separately running thread that will listen
+/// for incoming messages and handle them according to custom message handler.
+pub struct MessageChannel<H: ?Sized + MessageHandler + Send> {
+    publication_sender: UnboundedSender<ReceivedPublication>,
+    publication_receiver: UnboundedReceiver<ReceivedPublication>,
+    shutdown_handle: MessageHandlerShutdownHandle,
+    shutdown_recv: Receiver<()>,
+    message_handler: Box<H>,
+}
+
+impl<H> MessageChannel<H>
+where
+    H: MessageHandler + ?Sized + Send,
+{
+    pub fn new(message_handler: Box<H>) -> Self {
```

Comment:
nit: box it yourself
```suggestion
    pub fn new(message_handler: H) -> Self {
```

Reply:
We need to match on settings and return a trait object, which I believe needs to be a box. Therefore I think it is cleaner to pass it directly into this function as Box<> rather than unbox and rebox. What do you think?